### PR TITLE
[Ticket/12426] Allow search by custom profile fields

### PR DESCRIPTION
**Pull request by satanasov**
Checklist:

- [ ] Correct branch: master for new features; 3.3.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [ ] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket (set the ticket ID to 12426):

https://tracker.phpbb.com/browse/PHPBB3-12345

### CODE REVIEW

File: phpBB/cache/index.htm [hunk 1]
```diff
@@ -1,10 +0,0 @@
-<html>
```
Comment (3D-I) [inline phpBB/cache/index.htm:1]: Why has been deleted?

File: phpBB/memberlist.php [hunk 2]
```diff
@@ -1060,7 +1060,8 @@
 
 		// Additional sorting options for user search ... if search is enabled, if not
 		// then only admins can make use of this (for ACP functionality)
-		$sql_select = $sql_where_data = $sql_from = $sql_where = $order_by = '';
+		$sql_select = $sql_where_data = $sql_where = $order_by = '';
+		$sql_from = array();
```
Comment (3D-I) [inline phpBB/memberlist.php:1064]: pls use short array syntax `[]`

File: phpBB/memberlist.php [hunk 3]
```diff
@@ -1155,7 +1183,9 @@
 
 			if ($search_group_id)
 			{
-				$sql_from = ', ' . USER_GROUP_TABLE . ' ug ';
+				$sql_from = array(
```
Comment (3D-I) [inline phpBB/memberlist.php:1186]: pls use short array syntax `[]`

File: phpBB/memberlist.php [hunk 4]
```diff
@@ -1351,11 +1381,19 @@
 			);
 
 			$sql_select = ', ug.group_leader';
-			$sql_from = ', ' . USER_GROUP_TABLE . ' ug ';
+			$sql_join[] = array(
```
Comment (3D-I) [inline phpBB/memberlist.php:1384]: pls use short array syntax `[]`

File: phpBB/memberlist.php [hunk 5]
```diff
@@ -1399,10 +1437,27 @@
 		extract($phpbb_dispatcher->trigger_event('core.memberlist_modify_sql_query_data', compact($vars)));
 
 		// Count the users ...
-		$sql = 'SELECT COUNT(u.user_id) AS total_users
-			FROM ' . USERS_TABLE . " u$sql_from
-			WHERE " . $db->sql_in_set('u.user_type', $user_types) . "
-			$sql_where";
+		$sql_array = array(
```
Comment (3D-I) [inline phpBB/memberlist.php:1440]: pls use short array syntax `[]`

File: phpBB/memberlist.php [hunk 6]
```diff
@@ -1601,12 +1660,29 @@
 		$start = $pagination->validate_start($start, $config['topics_per_page'], $total_users);
 
 		// Get us some users :D
-		$sql = "SELECT u.user_id
-			FROM " . USERS_TABLE . " u
-				$sql_from
-			WHERE " . $db->sql_in_set('u.user_type', $user_types) . "
-				$sql_where
-			ORDER BY $order_by";
+		$sql_array = array(
```
Comment (3D-I) [inline phpBB/memberlist.php:1663]: pls use short array syntax `[]`

File: phpBB/phpbb/profilefields/manager.php [hunk 7]
```diff
@@ -720,5 +720,113 @@ private function list_profilefields($type_name, $active = false)
 		$this->db->sql_freeresult($result);
 
 		return array_column($rowset, 'field_ident', 'field_id');
+	}
+	/**
+	* Check if there is any record in CUSTOM_PROFILE_FILEDS_DATA
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/manager.php:725]: typo

File: phpBB/phpbb/profilefields/manager.php [hunk 8]
```diff
@@ -720,5 +720,113 @@ private function list_profilefields($type_name, $active = false)
...
+		return $data_exists;
+	}
+
+	/**
+	* Build Array for user search filter extension
+	* @return array
+	*/
+	public function build_custom_fields_search_array()
+	{
+		$custom_search_array = array();
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/manager.php:745]: pls use short array syntax `[]`

File: phpBB/phpbb/profilefields/type/type_base.php [hunk 9]
```diff
@@ -203,4 +203,22 @@ public function process_field_row($mode, $profile_row)
...
 		return $this->template->assign_display('cp_body');
 	}
+
+	/**
+	* Return templated value/field for search
+	*/
+	public function process_search_field_row($profile_row)
+	{
+		// set template filename
+		$this->template->set_filenames(array(
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/type/type_base.php:213]: pls use short array syntax `[]` - all over ;)

**Review by 3D-I — CHANGES_REQUESTED** [review]
First read - beside that there are also `phpBB > sniff` issues.

https://github.com/phpbb/phpbb/pull/6287/checks?check_run_id=3540084770

File: phpBB/cache/index.htm (hunk 1 above)
Comment (satanasov) [reply phpBB/cache/index.htm:1]: I don't know really ... apparently by mistake.

File: phpBB/memberlist.php [hunk 10]
```diff
@@ -1726,30 +1726,22 @@
...
 			if ($mode == 'group')
 			{
-				$sql_from_ary = explode(',', $sql_from);
-				$extra_tables = [];
-				foreach ($sql_from_ary as $entry)
-				{
-					$table_data = explode(' ', trim($entry));
-
-					if (empty($table_data[0]) || empty($table_data[1]))
+				$sql_array = array(
```
Comment (3D-I) [inline phpBB/memberlist.php:1729]: pls use short array syntax `[]`

File: phpBB/phpbb/profilefields/manager.php [hunk 11]
```diff
@@ -720,10 +720,9 @@ private function list_profilefields($type_name, $active = false)
 		$this->db->sql_freeresult($result);
 
 		return array_column($rowset, 'field_ident', 'field_id');
-	}
+	}
 	/**
-	* Check if there is any record in CUSTOM_PROFILE_FILEDS_DATA
-	* so we have any requests
+	* Check for data in CUSTOM_PROFILE_FILEDS_DATA
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/manager.php:725]: typo

should be `CUSTOM_PROFILE_FIELDS_DATA`

File: phpBB/phpbb/profilefields/manager.php [hunk 12]
```diff
@@ -722,7 +722,7 @@ private function list_profilefields($type_name, $active = false)
 		return array_column($rowset, 'field_ident', 'field_id');
 	}
 	/**
-	* Check for data in CUSTOM_PROFILE_FILEDS_DATA
+	* Check for data in CUSTOM_PROFILE_FIELDS_DATA
 	* @return boolean
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/manager.php:726]: Type return should be `bool`

File: phpBB/memberlist.php [hunk 13]
```diff
@@ -1601,12 +1660,29 @@
...
+			'WHERE'	=> $db->sql_in_set('u.user_type', $user_types) . $sql_where,
+			'ORDER_BY'	=> $order_by,
+		];
+
+		if ($config['load_cpf_memberlist'] && $additional_cpf_exist)
+		{
+			$sql_array['FROM'][PROFILE_FIELDS_DATA_TABLE] = 'pd';
+			$sql_array['WHERE'] .= ' AND u.user_id = pd.user_id';
+		}
+		if (!empty($sql_from))
```
Comment (DavidIQ) [inline phpBB/memberlist.php:1677]: `count` is the typical way we check for arrays having any elements.

File: phpBB/memberlist.php [hunk 14]
```diff
@@ -1351,11 +1381,19 @@
...
-			$sql_from = ', ' . USER_GROUP_TABLE . ' ug ';
+			$sql_join[] = [
+				'FROM'	=> [
+					USER_GROUP_TABLE	=> 'ug',
+				],
+				'ON' => 'u.user_id = ug.user_id',
+			];
+			//$sql_from[] = array(
+			//	USER_GROUP_TABLE	=> 'ug'
+			//);
```
[the remark this answers is not in the document]
Comment (3D-I) [reply phpBB/memberlist.php:1392]: IMHO those three commented lines should be removed.

File: phpBB/phpbb/profilefields/type/type_date.php [hunk 15]
```diff
@@ -273,6 +273,36 @@ public function generate_field($profile_row, $preview_options = false)
...
+	{
+		return '';
+	}
+
+	/**
+	* {@inheritDoc}
+	*/
+	public function get_search_array($profile_row)
+	{
+		$output = array(
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/type/type_date.php:289]: Short array syntax pls

File: phpBB/phpbb/profilefields/type/type_dropdown.php [hunk 16]
```diff
@@ -241,6 +241,69 @@ public function generate_field($profile_row, $preview_options = false)
...
+		}
+
+		$profile_row['field_value'] = (int) $value;
+		$this->template->assign_block_vars('dropdown', array_change_key_case($profile_row, CASE_UPPER));
+
+		$options = $this->lang_helper->get($profile_row['field_id'], $profile_row['lang_id']);
+
+		foreach ($options as $option_id => $option_value)
+		{
+			$this->template->assign_block_vars('dropdown.options', array(
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/type/type_dropdown.php:267]: Short array syntax pls

File: phpBB/phpbb/profilefields/type/type_dropdown.php [hunk 17]
```diff
@@ -241,6 +241,69 @@ public function generate_field($profile_row, $preview_options = false)
...
+			));
+		}
+	}
+
+	/**
+	* {@inheritDoc}
+	*/
+	public function get_search_array($profile_row)
+	{
+		$output = array(
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/type/type_dropdown.php:280]: Short array syntax pls

File: phpBB/phpbb/profilefields/type/type_int.php [hunk 18]
```diff
@@ -197,6 +197,60 @@ public function generate_field($profile_row, $preview_options = false)
...
+
+		$this->template->assign_block_vars('int', array_change_key_case($profile_row, CASE_UPPER));
+	}
+
+	/**
+	* {@inheritDoc}
+	*/
+	public function get_search_array($profile_row)
+	{
+		$output = array(
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/type/type_int.php:227]: Short array syntax pls

**Comment by 3D-I** [comment]
@satanasov you need to fix this commit messages _(use the usual commit template)_

image


image


---

And here, where `Ticket` should be `ticket`

image

File: phpBB/phpbb/profilefields/type/type_dropdown.php [hunk 19]
```diff
@@ -69,7 +69,6 @@ public function get_name_short()
 	public function get_options($default_lang_id, $field_data)
 	{
 		$profile_row = [];
-
 		$profile_row[0] = array(
```
Comment (3D-I) [inline phpBB/phpbb/profilefields/type/type_dropdown.php:72]: should be `$profile_row[0] = [`

Comment (satanasov) [reply phpBB/phpbb/profilefields/type/type_dropdown.php]: Merge resolution issue. Fixed.

**Comment by satanasov** [comment]
Closing for the moment ... will redo in a bit.

**Pull request closed without merging**